consensus/src/main/java/org/apache/iotdb/consensus/multileader/logdispatcher/LogDispatcher.java

```diff
@@ -0,0 +1,284 @@
+      this.peer = peer;
+      this.controller =
+          new IndexController(
+              impl.getStorageDir(), Utils.fromTEndPointToString(peer.getEndpoint()), false);
+      this.syncStatus = new SyncStatus(controller);
+    }
+
+    public IndexController getController() {
+      return controller;
+    }
+
+    public long getCurrentSyncIndex() {
+      return controller.getCurrentIndex();
+    }
+
+    public Peer getPeer() {
+      return peer;
+    }
+
+    public boolean offer(IndexedConsensusRequest request) {
+      return pendingRequest.offer(request);
+    }
+
+    public void stop() {
+      stopped = true;
+    }
+
+    @Override
+    public void run() {
+      logger.info("{}: Dispatcher for {} starts", impl.getThisNode(), peer);
+      try {
+        PendingBatch batch;
+        while (!Thread.interrupted() && !stopped) {
+          while ((batch = getBatch()).isEmpty()) {
+            // we may block here if there is no requests in the queue
+            bufferedRequest.add(pendingRequest.take());
+            // If write pressure is low, we simply sleep a little to reduce the number of RPC
+            if (pendingRequest.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+              Thread.sleep(MultiLeaderConsensusConfig.MAX_WAITING_TIME_FOR_ACCUMULATE_BATCH_IN_MS);
+            }
+          }
+          // we may block here if the synchronization pipeline is full
+          syncStatus.addNextBatch(batch);
+          // sends batch asynchronously and migrates the retry logic into the callback function
+          sendBatchAsync(batch, new DispatchLogHandler(this, batch));
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      } catch (Exception e) {
+        logger.error("Unexpected error in logDispatcher for peer {}", peer, e);
+      }
+      logger.info("{}: Dispatcher for {} exits", impl.getThisNode(), peer);
+    }
+
+    public PendingBatch getBatch() {
+      PendingBatch batch;
+      List<TLogBatch> logBatches = new ArrayList<>();
+      long startIndex = syncStatus.getNextSendingIndex();
+      long maxIndex = impl.getController().getCurrentIndex();
+      long endIndex;
+      if (bufferedRequest.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+        // Use drainTo instead of poll to reduce lock overhead
+        pendingRequest.drainTo(
+            bufferedRequest,
+            MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH - bufferedRequest.size());
+      }
+      if (bufferedRequest.isEmpty()) {
+        // only execute this after a restart
+        endIndex = constructBatchFromWAL(startIndex, maxIndex, logBatches);
+        batch = new PendingBatch(startIndex, endIndex, logBatches);
+        logger.debug("accumulated a {} from wal", batch);
+      } else {
+        Iterator<IndexedConsensusRequest> iterator = bufferedRequest.iterator();
+        IndexedConsensusRequest prev = iterator.next();
```

Comment:
In the scenario prev.index == 100, start = 1 and batch size = 10
We construct a batch from wal whose index rang is 2 - 11，and then we call constructBatchIndexedFromConsensusRequest and add the prev whose index is 100. I think this will cause an error.

Reply:
same as above